envision/server.py

```diff
@@ -54,30 +64,214 @@ def options(self):
+            sizes[idx_to_delete] = 0
+            self._frames[idx_to_delete].data = None
+
+
+class WebClientRunLoop:
+    """The run loop is like a "video player" for the simulation. It supports seeking
+    and playback. The run loop wraps the web client handler and pushes the frame
+    messages to it as needed.
+    """
+
+    def __init__(self, frames, web_client_handler, timestep_sec, seek=None):
+        self._frames = frames
+        self._client = web_client_handler
+        self._timestep_sec = timestep_sec
+        self._seek = seek
+        self._thread = None
+
+    def seek(self, offset_seconds):
+        self._seek = offset_seconds
+
+    def stop(self):
+        if self._thread:
+            self._thread.join(timeout=3)
+            self._thread = None
+
+    def run_forever(self):
+        async def run_loop():
+            # Wait till the first frame is present
+            frame = self._frames.start_frame
+            while frame is None:
+                await asyncio.sleep(0.1)
+                frame = self._frames.start_frame
+
+            while True:
+                # Handle seek
+                if self._seek is not None and self._frames.start_time is not None:
+                    frame = self._frames(self._frames.start_time + self._seek)
+                    frame = self._frame_with_data(frame)
+                    self._seek = None
+
+                if frame is None:
+                    # TODO: This shouldn't happen, log a warning
+                    frame = self._frames.start_frame
+
+                closed = self._push_frame_to_web_client(frame)
+                if closed:
+                    # XXX: Log warning that the socket closed and we're exiting
+                    return
+
+                frame = await self._wait_for_next_frame(frame)
+
+        def sync_run_forever():
+            loop = asyncio.new_event_loop()
+            asyncio.set_event_loop(loop)
+
+            loop.run_until_complete(run_loop())
+            loop.close()
+
+        self._thread = threading.Thread(target=sync_run_forever, args=(), daemon=True)
+        self._thread.start()
+
+    def _push_frame_to_web_client(self, frame):
+        try:
+            self._client.write_message(
+                json.dumps(
+                    {
+                        "state": frame.data,
+                        "current_elapsed_time": frame.timestamp
+                        - self._frames.start_time,
+                        "total_elapsed_time": self._frames.elapsed_time,
+                    }
+                )
+            )
+            return False
+        except WebSocketClosedError:
+            return True
+
+    async def _wait_for_next_frame(self, frame):
+        while True:
+            await asyncio.sleep(self._timestep_sec)
+            next_frame = self._frame_with_data(frame.next_)
+            if next_frame is not None:
+                return next_frame
+
+    def _frame_with_data(self, frame):
```

Comment:
After making the change above to completely remove a frame, you won't need this function any more